vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowSetImpl.java

```diff
@@ -42,43 +40,90 @@ static <U> Collector<Row, RowSetImpl<U>, RowSet<U>> collector(Function<Row, U> m
     return Collector.of(
       RowSetImpl::new,
       (set, row) -> {
-        set.list.add(mapper.apply(row));
+        set.add(mapper.apply(row));
       },
       (set1, set2) -> null, // Shall not be invoked as this is sequential
       (set) -> set
     );
   }
 
-  static Function<RowSet<Row>, RowSetImpl<Row>> FACTORY = rs -> (RowSetImpl) rs;
+  static Function<RowSet<Row>, RowSetImpl<Row>> FACTORY = rs -> (RowSetImpl<Row>) rs;
 
   static <U> Function<RowSet<U>, RowSetImpl<U>> factory() {
-    return rs -> (RowSetImpl) rs;
-  };
+    return rs -> (RowSetImpl<U>) rs;
+  }
 
-  private ArrayList<R> list = new ArrayList<>();
+  private R firstRow;
+  private Accumulator<R> rowAccumulator;
 
   @Override
   public RowSet<R> value() {
     return this;
   }
 
+  private void add(R row) {
+    if (rowAccumulator != null) {
+      rowAccumulator.accept(row);
+    } else if (firstRow != null) {
+      rowAccumulator = new ChunkedAccumulator<>(IntUnaryOperator.identity());
+      rowAccumulator.accept(firstRow);
+      rowAccumulator.accept(row);
+      firstRow = null;
+    } else {
+      firstRow = row;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
   @Override
   public RowIterator<R> iterator() {
-    Iterator<R> i = list.iterator();
-    return new RowIterator<R>() {
+    if (rowAccumulator != null) {
+      return rowAccumulator.iterator();
```

Comment:
Three different types here is bad for the JIT: better 2 and `null` if empty or just 2  (choosing the most common ones).

Reply:
ok